### ISSUE TITLE
Style/IdenticalConditionalBranches is incorrect when conditional statement has side effects.

### ISSUE THREAD

**Issue by jasonl** [rubocop-hq/rubocop#7651]
Given the following example, Rubocop's `Style/IdenticalConditionalBranches` flags that the repeated statement should be hoisted out of the conditional. But to do so would change the logic of the program.

```ruby
# frozen_string_literal: true

class SideEffectClass
  attr_reader :animal

  def should_do_something?
    @animal = rand(100) > 50 ? 'Dog' : 'Cat'
    rand(100) > 50
  end
end

side_effect = SideEffectClass.new

if side_effect.should_do_something?
  animal = side_effect.animal
  puts "#{animal} should do something."
else
  animal = side_effect.animal
  puts "#{animal} should not do something."
end
```

--------

## Expected behavior

Rubocop should not flag this code.

## Actual behavior

```
rubocop .
Inspecting 1 file
C

Offenses:

error.rb:3:1: C: Style/Documentation: Missing top-level class documentation comment.
class SideEffectClass
^^^^^
error.rb:15:3: C: Style/IdenticalConditionalBranches: Move animal = side_effect.animal out of the conditional.
  animal = side_effect.animal
  ^^^^^^^^^^^^^^^^^^^^^^^^^^^
error.rb:18:3: C: Style/IdenticalConditionalBranches: Move animal = side_effect.animal out of the conditional.
  animal = side_effect.animal
  ^^^^^^^^^^^^^^^^^^^^^^^^^^^

1 file inspected, 3 offenses detected
```

## Steps to reproduce the problem

See above

## RuboCop version

```
$ rubocop -v
0.79.0
```

**Comment by rrosenblum** [comment]
Unfortunately, there is no way for `Style/IdenticalConditionalBranches` to know if the code contains side effects. There is still an argument to be made that the structure of the code can/should still be changed to help readability.

The example that you provided can be rewritten as the following to still accomplish the same functionality.
```ruby
do_something = side_effect.should_do_something?
animal = side_effect.animal
if do_something
  puts "#{animal} should do something."
else
  puts "#{animal} should not do something."
end
```
